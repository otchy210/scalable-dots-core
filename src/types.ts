export type DotProps = {
  x: number;
  y: number;
  size: number;
  fill: string;
};

export type DotType = 'SQUARE' | 'CIRCLE' | 'RHOMBUS';

export type ImageData = {
  width: number;
  height: number;
  data: Uint8ClampedArray; // RGBA format, left to right then top to bottom.
};

export type ScalableDotsProps = {
  imageData: ImageData;
  type: DotType;
  size: number;
  gap: number;
};
