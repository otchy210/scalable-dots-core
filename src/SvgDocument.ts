import { DotProps, DotType } from './types';
import { XmlNode } from './XmlNode';

type SvgProps = {
  width: number;
  height: number;
};

export class SvgDocument {
  private root: XmlNode;

  constructor({ width, height }: SvgProps) {
    this.root = XmlNode.builder('svg')
      .attr('viewBox', `0 0 ${width} ${height}`)
      .attr('xmlns', 'http://www.w3.org/2000/svg')
      .attr('version', '1.1')
      .build();
  }
  addDot(dotType: DotType, dotProps: DotProps) {
    switch (dotType) {
      case 'SQUARE':
        this.addSquareDot(dotProps);
        break;
      case 'CIRCLE':
        this.addCircleDot(dotProps);
        break;
      case 'RHOMBUS':
        this.addRhombusDot(dotProps);
        break;
    }
  }
  addSquareDot({ x, y, size, fill }: DotProps) {
    const node = XmlNode.builder('path')
      .attr('d', `m${x} ${y}h${size}v${size}h-${size}z`)
      .attr('fill', fill)
      .build();
    this.root.addChild(node);
  }
  addCircleDot({ x, y, size, fill }: DotProps) {
    const r = size / 2;
    const cx = x + r;
    const cy = y + r;
    const node = XmlNode.builder('circle')
      .attr('cx', cx)
      .attr('cy', cy)
      .attr('r', r)
      .attr('fill', fill)
      .build();
    this.root.addChild(node);
  }
  addRhombusDot({ x, y, size, fill }: DotProps) {
    const hs = size / 2;
    const node = XmlNode.builder('path')
      .attr('d', `m${x} ${y + hs}l${hs} -${hs}l${hs} ${hs}l-${hs} ${hs}z`)
      .attr('fill', fill)
      .build();
    this.root.addChild(node);
  }
  toMinifiedXml() {
    return this.root.toMinifiedXml();
  }
  toPrettyXml(tab = '  ') {
    return this.root.toPrettyXml(tab);
  }
}
