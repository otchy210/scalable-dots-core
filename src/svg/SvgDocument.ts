import { XmlNode } from './XmlNode';

type SvgProps = {
  width: number;
  height: number;
};

type DotProps = {
  x: number;
  y: number;
  size: number;
  fill: string;
};

export class SvgDocument {
  private root: XmlNode;

  constructor({ width, height }: SvgProps) {
    this.root = XmlNode.builder('svg')
      .attr('viewBox', `0 0 ${width} ${height}`)
      .attr('xmlns', 'http://www.w3.org/2000/svg')
      .attr('version', '1.1')
      .build();
  }
  addSquareDot({ x, y, size, fill }: DotProps) {
    const node = XmlNode.builder('path')
      .attr('d', `m${x} ${y}h${size}v${size}h-${size}z`)
      .attr('fill', fill)
      .build();
    this.root.addChild(node);
  }
  addCircleDot({ x, y, size, fill }: DotProps) {
    const r = size / 2;
    const cx = x + r;
    const cy = y + r;
    const node = XmlNode.builder('circle')
      .attr('cx', cx)
      .attr('cy', cy)
      .attr('r', r)
      .attr('fill', fill)
      .build();
    this.root.addChild(node);
  }
  toMinifiedXml() {
    return this.root.toMinifiedXml();
  }
  toPrettyXml() {
    return this.root.toPrettyXml();
  }
}
