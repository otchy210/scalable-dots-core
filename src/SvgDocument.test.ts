import { SvgDocument } from './SvgDocument';
import { DotProps, DotType } from './types';

describe('SvgDocument', () => {
  it('renders empty svg properly', () => {
    const svg = new SvgDocument({ width: 50, height: 100 });
    expect(svg.toMinifiedXml()).toMatchSnapshot();
    expect(svg.toPrettyXml()).toMatchSnapshot();
  });
  describe('addSquareDot', () => {
    it('works', () => {
      const svg = new SvgDocument({ width: 20, height: 20 });
      svg.addSquareDot({ x: 0, y: 0, size: 10, fill: 'red' });
      svg.addSquareDot({ x: 0, y: 10, size: 10, fill: 'green' });
      svg.addSquareDot({ x: 10, y: 0, size: 10, fill: 'blue' });
      svg.addSquareDot({ x: 10, y: 10, size: 10, fill: 'yellow' });
      expect(svg.toPrettyXml()).toMatchSnapshot();
    });
  });
  describe('addCircleDot', () => {
    it('works', () => {
      const svg = new SvgDocument({ width: 20, height: 20 });
      svg.addCircleDot({ x: 0, y: 0, size: 10, fill: 'red' });
      svg.addCircleDot({ x: 0, y: 10, size: 10, fill: 'green' });
      svg.addCircleDot({ x: 10, y: 0, size: 10, fill: 'blue' });
      svg.addCircleDot({ x: 10, y: 10, size: 10, fill: 'yellow' });
      expect(svg.toPrettyXml()).toMatchSnapshot();
    });
  });
  describe('addRhombusDot', () => {
    it('works', () => {
      const svg = new SvgDocument({ width: 20, height: 20 });
      svg.addRhombusDot({ x: 0, y: 0, size: 10, fill: 'red' });
      svg.addRhombusDot({ x: 0, y: 10, size: 10, fill: 'green' });
      svg.addRhombusDot({ x: 10, y: 0, size: 10, fill: 'blue' });
      svg.addRhombusDot({ x: 10, y: 10, size: 10, fill: 'yellow' });
      expect(svg.toPrettyXml()).toMatchSnapshot();
    });
  });
  describe('addDot', () => {
    type TestSig = [type: DotType, result: string];
    const dotProps: DotProps = {
      x: 0,
      y: 0,
      size: 10,
      fill: '#000',
    };
    it.each<TestSig>([
      ['SQUARE', 'm0 0h10v10h-10z'],
      ['CIRCLE', 'r="5"'],
      ['RHOMBUS', 'm0 5l5 -5l5 5l-5 5z'],
    ])('adds proper %s', (type, result) => {
      const svg = new SvgDocument({ width: 10, height: 10 });
      svg.addDot(type, dotProps);
      expect(svg.toMinifiedXml().includes(result)).toBe(true);
    });
  });
});
