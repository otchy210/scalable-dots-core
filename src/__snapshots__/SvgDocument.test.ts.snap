// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SvgDocument addCircleDot works 1`] = `
"<svg viewBox=\\"0 0 20 20\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\">
  <circle cx=\\"5\\" cy=\\"5\\" r=\\"5\\" fill=\\"red\\"/>
  <circle cx=\\"5\\" cy=\\"15\\" r=\\"5\\" fill=\\"green\\"/>
  <circle cx=\\"15\\" cy=\\"5\\" r=\\"5\\" fill=\\"blue\\"/>
  <circle cx=\\"15\\" cy=\\"15\\" r=\\"5\\" fill=\\"yellow\\"/>
</svg>"
`;

exports[`SvgDocument addRhombusDot works 1`] = `
"<svg viewBox=\\"0 0 20 20\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\">
  <path d=\\"m0 5l5 -5l5 5l-5 5z\\" fill=\\"red\\"/>
  <path d=\\"m0 15l5 -5l5 5l-5 5z\\" fill=\\"green\\"/>
  <path d=\\"m10 5l5 -5l5 5l-5 5z\\" fill=\\"blue\\"/>
  <path d=\\"m10 15l5 -5l5 5l-5 5z\\" fill=\\"yellow\\"/>
</svg>"
`;

exports[`SvgDocument addSquareDot works 1`] = `
"<svg viewBox=\\"0 0 20 20\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\">
  <path d=\\"m0 0h10v10h-10z\\" fill=\\"red\\"/>
  <path d=\\"m0 10h10v10h-10z\\" fill=\\"green\\"/>
  <path d=\\"m10 0h10v10h-10z\\" fill=\\"blue\\"/>
  <path d=\\"m10 10h10v10h-10z\\" fill=\\"yellow\\"/>
</svg>"
`;

exports[`SvgDocument renders empty svg properly 1`] = `"<svg viewBox=\\"0 0 50 100\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\"/>"`;

exports[`SvgDocument renders empty svg properly 2`] = `"<svg viewBox=\\"0 0 50 100\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\"/>"`;
